src/array.jl

```diff
@@ -479,20 +482,76 @@ similar(A::CategoricalArray{S, M, R}, ::Type{T},
 similar(A::CategoricalArray{S, M, R}, ::Type{Missing},
         dims::NTuple{N, Int}) where {N, S, M, R} =
     Array{Missing, N}(missing, dims)
-similar(A::CategoricalArray{S, M, Q}, ::Type{T},
-        dims::NTuple{N, Int}) where {R, T<:CatValue{R}, N, S, M, Q} =
+similar(A::CategoricalArray{S, M, Q}, ::Type{CategoricalValue{T, R}},
+        dims::NTuple{N, Int}) where {R, T, N, S, M, Q} =
     CategoricalArray{T, N, R}(undef, dims)
-similar(A::CategoricalArray{S, M, R}, ::Type{T},
-        dims::NTuple{N, Int}) where {S, T<:CatValue, M, N, R} =
-    CategoricalArray{T, N, R}(undef, dims)
-# Union{T, Missing} is repeated even if theoretically redundant because of JuliaLang/julia#26405
-# Once that bug is fixed, Union{T, Missing} can be replaced with T and the two definitions above can be removed
-similar(A::CategoricalArray{S, M, Q}, ::Type{T},
-        dims::NTuple{N, Int}) where {R, T<:Union{CatValue{R}, Missing}, N, S, M, Q} =
-    CategoricalArray{Union{T, Missing}, N, R}(undef, dims)
-similar(A::CategoricalArray{S, M, R}, ::Type{T},
-        dims::NTuple{N, Int}) where {S, T<:Union{CatValue, Missing}, M, N, R} =
+similar(A::CategoricalArray{S, M, Q}, ::Type{CategoricalValue{T}},
+        dims::NTuple{N, Int}) where {T, N, S, M, Q} =
+    CategoricalArray{T, N}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{CategoricalString{R}},
+        dims::NTuple{N, Int}) where {R, N, S, M, Q} =
+    CategoricalArray{String, N, R}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{CategoricalString},
+        dims::NTuple{N, Int}) where {N, S, M, Q} =
+    CategoricalArray{String, N, Q}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{Union{CategoricalValue{T, R}, Missing}},
+        dims::NTuple{N, Int}) where {R, T, N, S, M, Q} =
     CategoricalArray{Union{T, Missing}, N, R}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{Union{CategoricalValue{T}, Missing}},
+        dims::NTuple{N, Int}) where {T, N, S, M, Q} =
+    CategoricalArray{Union{T, Missing}, N, Q}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{Union{CategoricalString{R}, Missing}},
+        dims::NTuple{N, Int}) where {R, N, S, M, Q} =
+    CategoricalArray{Union{String, Missing}, N, R}(undef, dims)
+similar(A::CategoricalArray{S, M, Q}, ::Type{Union{CategoricalString, Missing}},
+        dims::NTuple{N, Int}) where {N, S, M, Q} =
+    CategoricalArray{Union{String, Missing}, N, Q}(undef, dims)
+
+for A in (:AbstractArray, :Array, :Vector, :Matrix) # to fix ambiguities
```

Comment:
I thought you would end up using `@eval` :).
Actually I thought you will use for these 4 methods that I thought are enough to start with.

Reply:
Unfortunately all methods are needed AFAICT, since broadcasting calls `similar` both with an array type and with an array instance, and they aren't connected currently.

Reply:
I know - I just expected that we will get into things, like `@eval`, I typically try to avoid when writing code. But here it is clear that this is needed